### PR TITLE
Initial commit to support the specification of users in External Project Access scanner rule.

### PR DESCRIPTION
**Pull request by kevensen**
For the external project access scanner, this allows the specification of a list of users for which the rule applies.

Addresses #2525

### CODE REVIEW

File: tests/scanner/audit/external_project_access_engine_test.py [hunk 1]
```diff
@@ -162,16 +165,29 @@ def test_validate_bad_ancestor(self):
             self.rule_book.validate_ancestors(
                 self.TEST_BAD_RULE['allowed_ancestors'], 0)
 
+    def test_validate_good_user(self):
+        """Test proper rule validation"""
+        self.rule_book.validate_users(
```
Comment (blueandgold) [inline tests/scanner/audit/external_project_access_engine_test.py]: There should be an assert in this test.

File: tests/scanner/audit/external_project_access_engine_test.py [hunk 2]
```diff
@@ -211,27 +227,95 @@ class ExternalProjectAccessRuleTest(ForsetiTestCase):
...
         rule = engine_module.Rule(rule_name='test_single_item_in_rule_match',
                                   rule_index=0,
-                                  rules=[Organization('7890')])
+                                  rules=processed_rule)
+        violation = rule.find_violation('user1@example.com',
+                                        self.TEST_ANCESTORS)
+        self.assertIsNone(violation)
+
+    def test_single_item_in_rule_match_with_user(self):
+        """Test no violations are found with single item in rule"""
```
Comment (blueandgold) [inline tests/scanner/audit/external_project_access_engine_test.py]: This description should be a bit different for this test, i.e. add "with user".  Can you please update?

File: tests/scanner/audit/external_project_access_engine_test.py [hunk 3]
```diff
@@ -211,27 +227,95 @@ class ExternalProjectAccessRuleTest(ForsetiTestCase):
...
+
+        rule = engine_module.Rule(rule_name='test_single_item_in_rule_match_with_user',
+                                  rule_index=0,
+                                  rules=processed_rule)
         violation = rule.find_violation('user1@example.com',
                                         self.TEST_ANCESTORS)
         self.assertIsNone(violation)
 
+    def test_single_item_in_rule_match_with_wrong_user(self):
+        """Test no violations are found with single item in rule"""
```
Comment (blueandgold) [inline tests/scanner/audit/external_project_access_engine_test.py]: Ditto, on making this description more specific for this test.

File: tests/scanner/audit/external_project_access_engine_test.py [hunk 4]
```diff
@@ -211,27 +227,95 @@ class ExternalProjectAccessRuleTest(ForsetiTestCase):
...
         rule = engine_module.Rule(rule_name='test_multi_items_in_rule_match',
                                   rule_index=0,
-                                  rules=[Folder('456'), Organization('7890')])
+                                  rules=processed_rule)
         violation = rule.find_violation('user1@example.com',
                                         self.TEST_ANCESTORS)
         self.assertIsNone(violation)
 
+    def test_multi_items_in_rule_match_with_user(self):
+        """Test no violations are found with multiple items in rule"""
```
Comment (blueandgold) [inline tests/scanner/audit/external_project_access_engine_test.py]: ditto on description

File: tests/scanner/audit/external_project_access_engine_test.py [hunk 5]
```diff
@@ -211,27 +227,95 @@ class ExternalProjectAccessRuleTest(ForsetiTestCase):
...
+
+        rule = engine_module.Rule(rule_name='test_multi_items_in_rule_match_with_user',
+                                  rule_index=0,
+                                  rules=processed_rule)
+        violation = rule.find_violation('user1@example.com',
+                                        self.TEST_ANCESTORS)
+        self.assertIsNone(violation)
+
+    def test_multi_items_in_rule_match_with_wrong_user(self):
+        """Test no violations are found with multiple items in rule"""
```
Comment (blueandgold) [inline tests/scanner/audit/external_project_access_engine_test.py]: ditto on description

File: tests/scanner/audit/external_project_access_engine_test.py [hunk 6]
```diff
@@ -242,12 +326,55 @@ def test_single_item_no_match(self):
...
+
+        self.assertEqual(0, violation.rule_index)
+        self.assertEqual('test_single_item_no_match_with_user',
+                         violation.rule_name)
+        self.assertEqual('projects/123', violation.full_name)
+        self.assertEqual('projects/123,folders/456,organizations/7890',
+                         violation.resource_data)
+
+    def test_single_item_no_match_with_wrong_user(self):
+        """Test violations are found with single item in rule"""
```
Comment (blueandgold) [inline tests/scanner/audit/external_project_access_engine_test.py]: ditto on description, add "with wrong user"; can you please update other descriptions too?

File: google/cloud/forseti/scanner/audit/external_project_access_rules_engine.py [hunk 7]
```diff
@@ -82,6 +82,8 @@ class ExternalProjectAccessRuleBook(bre.BaseRuleBook):
 
     # Class variable for matching the ancestor during rule validation
     ancestor_pattern = re.compile(r'^organizations/\d+$|^folders/\d+$')
+    email_pattern = \
```
Comment (blueandgold) [inline google/cloud/forseti/scanner/audit/external_project_access_rules_engine.py]: line continuation should be with paranthesis
```
email_pattern = (
    re.compile(r'(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)')
```

**Comment by hshin-g** [comment]
@kevensen Can you update the rule file with documentation that outlines this fix?

**Pull request merged**